### ISSUE TITLE
Prism Snippets

### ISSUE THREAD

**Issue by mtaulbee** [PrismLibrary/Prism#330]
Hello,
After using Prism for the last month, I've created a few code snippets to help speed things up.
I've created the following.....
1. prismProp - This is basically the same thing as a normail full property, but it uses SetProperty method in the setter.
2. prismCommand - Creates a prism delegate command.
3. prismClass - Inherits from the BindableBase class in prism.

Could we include these snippets with Prism?
I'm also open to creating more snippets based on community feedback.

**Comment by brianlagunas** [comment]
I actually have the same prismprop snippet :)

Yes, I am heads down in feature work, but please copy/paste your snippets here and I will add them to the list that we want to release

**Comment by mtaulbee** [comment]
Any other snippets you would like to include?

**Comment by brianlagunas** [comment]
I will review what you have and see what makes sense to include when we finally get around to it.

**Comment by bartlannoeye** [comment]
I got a propp (aka prismprop) as well, but then as R# template.

We've spoken about it, but never came to releasing them. Maybe we can package them in a small vsix?

**Comment by mtaulbee** [comment]
@bartlannoeye 
I was thinking the same thing. It would be nice to include some templates as well, maybe include this with issue #13?

**Comment by mtaulbee** [comment]
The command snippets are similar to MVVM light. 
I've added summaries to these snippets, but I don't use them in my personal snippets.

[prismDelCmdCanExecute.txt](https://github.com/PrismLibrary/Prism/files/52047/prismDelCmdCanExecute.txt)
[prismProp.txt](https://github.com/PrismLibrary/Prism/files/52050/prismProp.txt)
[prismVM.txt](https://github.com/PrismLibrary/Prism/files/52049/prismVM.txt)
[prismDelCmd.txt](https://github.com/PrismLibrary/Prism/files/52048/prismDelCmd.txt)

**Comment by brianlagunas** [comment]
So far, the only one that I would probably put in is the prismprop.  Maybe the prismVM one, but I will have to play with it.  Basically you have to create a new file, then delete what is there and then use the snippet.  Maybe a new item template would be best in this case?

As for the command snippets, I would not provide that type of guidance for defining commands so these would not be included in the official prism VSIX.

**Comment by mtaulbee** [comment]
I agree with your statements. I think prismProp is a much needed snippet and should be included in the next release if possible.

**Comment by boodoolak** [comment]
I have long waited for this. Snippets are very necessary.

**Comment by brianlagunas** [comment]
The first version of this has been released.  Please test it out and let me know if you run into any issues:

https://visualstudiogallery.msdn.microsoft.com/e7b6bde2-ba59-43dd-9d14-58409940ffa0

**Comment by mtaulbee** [comment]
The prismprop snippet works as expected, great addition!

**Comment by brianlagunas** [comment]
great, now we just need to identify more snippets for code we write a the time, but also make sense to add to the plugin

**Comment by brianlagunas** [comment]
I'm curious if a PubSubEvent snippet would be better than a PubSubEvent item template, or maybe both?  Hmm... maybe a CompositeCommand definition snippet/item template too?

**Comment by mtaulbee** [comment]
I agree with PubSub, i use CompositeCommand, but I've never really felt a snippet was needed. That just might be me though....

**Comment by brianlagunas** [comment]
So this is what I am thinking of adding so far:

prismpropcmd - new DelegateCommand property (only getter/setter)
prismpropcmdgeneric new DelegateCommand of T generic property (only getter/setter)
prismpubsubevent - new PuSubEvent deinfition (should this also be an Itemtemplate or would a snippet do)

**Comment by boodoolak** [comment]
I can't find how to use ItemTemplates
Prism ViewModel - A ViewModel that derives from BindableBase and has a default constructor.
Menu "Add - New Item" not contain this template.

**Comment by mtaulbee** [comment]
Those work for me. The only request I have is that the setters for the commands are private.
If they aren't then I can tell you I'll never use them.
As far as the pubsub, I've never considered an item template. If it's because of the payload, I can't say the payloads I create are ever that "heavy" IE not a lot of properties. Maybe that's just me though....

**Comment by brianlagunas** [comment]
@boodoolak it's under the Prism category. You can also just search the template for Prism ViewModel.

**Comment by mtaulbee** [comment]
@boodoolak Please see attached image....
![prism vm template](https://cloud.githubusercontent.com/assets/15049743/11745463/5fdb287a-9fe6-11e5-990f-d7ca41fdd82c.png)

**Comment by boodoolak** [comment]
Aaah.. I have UWP project... Templates only for wpf and xamarin? I have not installed wpf and i have not this items in add menu window. 
![image](https://cloud.githubusercontent.com/assets/3179176/11752601/95b79fbc-a051-11e5-90bf-16dce7a32c74.png)

**Comment by brianlagunas** [comment]
@boodoolak We don't have any UWP templates yet.

**Comment by boodoolak** [comment]
Ok, I'll be waiting :-)

**Comment by descite** [comment]
Please add 
`<SnippetTypes>
    <SnippetType>Expansion</SnippetType>
</SnippetTypes>`
to snippets header. then snippets are usable directly from editor ( like prop snippet already available in vs).
And we are not needed to use ctrl+k,ctl+s to use snippet. (visual studio 2015).

**Comment by nguyenthanhliemfc** [comment]
@brianlagunas Hi! How to Inport your Prism Snippet code to Resharper. Thank you!

**Comment by brianlagunas** [comment]
@nguyenthanhliemfc You'll have to ask Resharper that question.  YOu can find the snippets here:  https://github.com/PrismLibrary/Prism-Extensibility/tree/master/TemplatePack/TemplatePack/Snippets/CSharp/Prism

Good luck.

**Comment by bartlannoeye** [comment]
@nguyenthanhliemfc you can still use Visual Studio snippets when using ReSharper. They work both side by side. If you want to use [ReSharper templates](https://www.jetbrains.com/resharper/features/code_templates.html) (that's what they are instead of snippets), you'll have to transform them a bit. I could maybe create a separate Visual Studio extension for these snippets, as I don't think we want them in the default Prism extension.

**Comment by nguyenthanhliemfc** [comment]
@brianlagunas  your snippets code link die!!!

**Comment by brianlagunas** [comment]
@nguyenthanhliemfc either VS is messing up, or you could have a conflict.  Easy to fix though.

Find the Prism Folder in Code Snippets Manager, and update the URI to this location:

C:\Users\{UserName}\AppData\Local\Microsoft\VisualStudio\(Version of VS)\Extensions\{Install ID}\Snippets\CSharp\Prism

You will have to browse and find the proper Install ID.

**Comment by andyfmking** [comment]
Yup, this seems like a problem with the folder path in the Tools -> Code Snippets Manager.

When you choose csharp from the dropdown  selection for the programming language specific snippets, you'l find wrong folder location for the Prism folder (has an exclamation mark)

Seems after installing the Prism Templates, wrong folder location gets added in the Snippets Manager pointing to a NON-Existing location.

**SOLUTION:**
Remove the wrong folder and "ADD" the folder from as described by Brian

**C:\Users{UserName}\AppData\Local\Microsoft\VisualStudio(Version of VS)\Extensions{Install ID}\Snippets\CSharp\Prism**

I had to change it manually to get the snippets to work.

Took me a good 30 minutes of searching for this solution till I saw this reply.

I'm not sure if installing Visual Studio 2017 in drive D:\ instead of C:\ could have cause this (just thinking out loud)
